instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java

```diff
@@ -91,96 +99,119 @@ Span nextSpan(TraceContextOrSamplingFlags extracted, GrpcServerRequest request)
       : tracer.nextSpan(extracted);
   }
 
+  void finish(GrpcServerResponse response, @Nullable Span span) {
+    if (span == null || span.isNoop()) return;
+    Throwable error = response.error();
+    if (error != null) span.error(error);
+    parser.onClose(response.status, response.trailers, span.customizer());
+    span.finish();
+  }
+
+  void finishWithError(@Nullable Span span, Throwable error) {
+    if (span == null || span.isNoop()) return;
+    if (error != null) span.error(error);
+    span.finish();
+  }
+
   final class TracingServerCall<ReqT, RespT> extends SimpleForwardingServerCall<ReqT, RespT> {
-    final Span span;
+    final TraceContext context;
+    final AtomicReference<Span> spanRef;
+    final GrpcServerRequest request;
 
-    TracingServerCall(Span span, ServerCall<ReqT, RespT> call) {
-      super(call);
-      this.span = span;
+    TracingServerCall(ServerCall<ReqT, RespT> delegate, Span span, AtomicReference<Span> spanRef,
+      GrpcServerRequest request) {
+      super(delegate);
+      this.context = span.context();
+      this.spanRef = spanRef;
+      this.request = request;
     }
 
     @Override public void request(int numMessages) {
-      span.start();
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        super.request(numMessages);
+      try (Scope scope = currentTraceContext.maybeScope(context)) {
+        delegate().request(numMessages);
       }
     }
 
     @Override public void sendHeaders(Metadata headers) {
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        super.sendHeaders(headers);
+      try (Scope scope = currentTraceContext.maybeScope(context)) {
+        delegate().sendHeaders(headers);
       }
     }
 
     @Override public void sendMessage(RespT message) {
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        super.sendMessage(message);
-        parser.onMessageSent(message, span.customizer());
+      try (Scope scope = currentTraceContext.maybeScope(context)) {
+        delegate().sendMessage(message);
+        Span span = spanRef.get(); // could be an error
+        SpanCustomizer customizer = span != null ? span.customizer() : NoopSpanCustomizer.INSTANCE;
+        messageProcessor.onMessageSent(message, customizer);
       }
     }
 
     @Override public void close(Status status, Metadata trailers) {
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        super.close(status, trailers);
-        parser.onClose(status, trailers, span.customizer());
+      Throwable error = null;
+      try (Scope scope = currentTraceContext.maybeScope(context)) {
+        delegate().close(status, trailers);
       } catch (Throwable e) {
-        span.error(e);
+        error = e;
```

Comment:
I'm less clear about this one since closure on the server side is handled by the framework, so an error may or may not indicate the response failed. Ideally, our interceptor is in a location where the only possibility of an exception here is a grpc-java bug.

Reply:
agreed and in rationale now. thanks!